README.md
# Jarvis

[![DOI](https://zenodo.org/badge/doi/10.5281/zenodo.7779363.png)](https://zenodo.org/badge/latestdoi/612752168)

Jarvis is an AI assistant for compiling scientific literature auto-reviews based on free text prompts. This extension was ported from [Jarvis](https://github.com/alondmnt/joplin-plugin-jarvis) for the [Joplin](https://joplinapp.org) note-taking app.

## Features

Write what you're interested about, select the text, and run the command "Research with Jarvis". Wait 2-3 minutes for all the output to appear in the note (depending on internet traffic). Jarvis will update the content as it finds new information on the web (using Semantic Scholar, Crossref, Elsevier, Springer & Wikipedia databases). In the end you will get a report with the following sections: title, prompt, research questions, queries, references, review and follow-up questions.

![](https://github.com/alondmnt/vscode-jarvis/blob/main/images/jarvis-demo.gif?raw=true)

## Dependencies

- For GPT-3, you will need to get a free [OpenAI API Key](https://platform.openai.com/account/api-keys)
- For searching via Scopus (optional, recommended to try) you will need to get a free [Scopus API key](https://dev.elsevier.com/apikey/manage)
- For access to Springer's paper repository (optional) you will need to get a free [Springer API key](https://dev.springernature.com/admin/applications)

## Release Notes

### 0.0.4

Improved token estimation.
License: AGPLv3.

### 0.0.2-0.0.3

Improved Wikipedia summary.

### 0.0.1

Ported Jarvis (research mode) from Joplin.

---


CHANGELOG.md
# Change Log

All notable changes to the Jarvis extension will be documented in this file.

## [Unreleased]

## [0.0.4]

- Improved token estimation
- License: AGPLv3

## [0.0.3]

### Improved

- Wikipedia summary

### Changed

- Default temperature: 1.0

### Fixed

- Prompt typos
- Handle empty search results

## [0.0.1]

### Added

- Ported Jarvis (research mode) from Joplin

[Unreleased]: https://github.com/alondmnt/vscode-jarvis/compare/v0.0.4...main
[0.0.4]: https://github.com/alondmnt/vscode-jarvis/compare/v0.0.3...v0.0.4
[0.0.3]: https://github.com/alondmnt/vscode-jarvis/compare/v0.0.1...v0.0.3
[0.0.1]: https://github.com/alondmnt/vscode-jarvis/releases/tag/v0.0.1


src/openai.ts
/* eslint-disable @typescript-eslint/naming-convention */
import * as vscode from 'vscode';
import axios from 'axios';
import { JarvisSettings } from './settings';

export async function query_completion(
    prompt: string, settings: JarvisSettings,
    adjust_max_tokens: number = 0): Promise<string> {
  if ( !settings.openai_api_key ) {
    vscode.window.showErrorMessage('OpenAI API key not set');
    throw new Error('OpenAI API key not set');
  }

  const input_tokens = Math.ceil(prompt.length / 4) + 20;  // heuristic for input length with some room for error
  let url: string = '';
  let responseParams: any = {
    model: settings.model,
    max_tokens: settings.max_tokens - input_tokens - adjust_max_tokens,
    temperature: settings.temperature,
    top_p: settings.top_p,
    frequency_penalty: settings.frequency_penalty,
    presence_penalty: settings.presence_penalty,
  };

  if (settings.model.includes('gpt-3.5-turbo')) {
    url = 'https://api.openai.com/v1/chat/completions';
    responseParams = {...responseParams,
      messages: [
        {role: 'system', content: 'You are Jarvis, the helpful assistant.'},
        {role: 'user', content: prompt}
      ],
    };
    responseParams['max_tokens'] -= 10;  // correction due to system message
  } else {
    url = 'https://api.openai.com/v1/completions';
    responseParams = {...responseParams,
      prompt: prompt,
    };
  }

  const response = await axios.post(url, responseParams, {
    timeout: settings.timeout,
    headers: {
      Authorization: 'Bearer ' + settings.openai_api_key
    }
  }).catch((error) => {
    console.log('OpenAI error: ', error);
    return error;
  });
  const data = response.data;

  if (response.message && response.message.includes('timeout')) {
    const selection = await vscode.window.showErrorMessage(
      'OpenAI request timed out. You can increase the timeout in the settings.',
      'Retry', 'Cancel');
    if ( selection === 'Cancel' ) {
      return '';
    } else {
      console.log('retrying request');
      return await query_completion(prompt, settings, adjust_max_tokens);
    }
  }
  if (response.status !== 200 && response.response.status !== 400) {
    vscode.window.showErrorMessage('OpenAI request failed: ' + response.error.message);
    return '';
  }

  if (data) {
    // output completion
    if (data.hasOwnProperty('choices') && (data.choices[0].text)) {
      return data.choices[0].text;
    }
    if (data.hasOwnProperty('choices') && data.choices[0].message.content) {
      return data.choices[0].message.content;
    }
  }

  // display error message
  const error = response.response.data.error.message;
  const selection = await vscode.window.showErrorMessage(
    `Error: ${error}`, 'Retry', 'Cancel');
  if ( selection === 'Cancel' ) {
    return '';
  }
  // adjust & retry

  // find all numbers in error message
  const max_tokens = [...error.matchAll(/([0-9]+)/g)];

  // adjust max tokens
  if ((max_tokens !== null) &&
      (error.includes('reduce'))) {
    adjust_max_tokens = parseInt(max_tokens[1]) - parseInt(max_tokens[0]) + 1;
  }

  // retry
  return await query_completion(prompt, settings, adjust_max_tokens);
}


src/research.ts
/* eslint-disable @typescript-eslint/naming-convention */
import * as vscode from 'vscode';
import { get_settings } from './settings';
import { query_completion } from './openai';
import { PaperInfo, SearchParams, search_papers, sample_and_summarize_papers } from './papers';
import { WikiInfo, search_wikipedia } from './wikipedia';
import { replace_selection } from './utils';

export async function do_research(prompt: string) {
  const settings = get_settings();
  const n_papers = settings.paper_space;
  const paper_tokens = Math.ceil(settings.paper_tokens / 100 * settings.max_tokens);
  const use_wikipedia = settings.use_wikipedia;

  let [papers, search] = await search_papers(prompt, n_papers, settings);
  console.log(`papers: ${papers.length}`);

  replace_selection(search.response);
  let wiki_search: Promise<WikiInfo> = Promise.resolve({ summary: '' });
  if ( use_wikipedia && (papers.length > 0) ) {
    // start search in parallel to paper summary
    wiki_search = search_wikipedia(prompt, search, settings);
  }
  papers = await sample_and_summarize_papers(papers, paper_tokens, search, settings);
  console.log(`papers sampled and summarized (${papers.length})`);

  if (papers.length === 0) {
    replace_selection('No relevant papers found. Consider expanding your paper space, resending your prompt, or adjusting it.\n');
    return;
  }

  const full_prompt = get_full_prompt(papers, await wiki_search, search);
  const research = await query_completion(full_prompt, settings);
  console.log(`research completed (${research.length} chars)`);
  replace_selection('\n## Review\n\n' + research.trim());
}

function get_full_prompt(papers: PaperInfo[], wiki: WikiInfo, search: SearchParams): string {
  let full_prompt = 
    `write a response to the prompt. address the research questions.
    use all relevant papers listed below, and cite what you use in the response.
    DO NOT cite papers other than the provided ones, but you may add additional uncited information that might be considered common knowledge.
    try to explain acronyms and definitions of domain-specific terms.
    finally, add a section of "## Follow-up questions" to the response.\n\n`;
  full_prompt += wiki['summary'] + '\n\n';
  for (let i = 0; i < papers.length; i++) {
    full_prompt += papers[i]['summary'] + '\n\n';
  }
  full_prompt += `## Prompt\n\n${search.prompt}\n`;
  full_prompt += `## Research questions\n\n${search.questions}\n`;
  return full_prompt;
}


src/settings.ts
/* eslint-disable @typescript-eslint/naming-convention */
import * as vscode from 'vscode';

export const search_prompts: { [engine: string] : string; } = {
  'Scopus': `
    next, generate a few valid Scopus search queries, based on the questions and prompt, using standard Scopus operators.
    try to use various search strategies in the multiple queries. for example, if asked to compare topics A and B, you could search for ("A" AND "B"),
    and you could also search for ("A" OR "B") and then compare the results.
    only if explicitly required in the prompt, you can use additional operators to filter the results, like the publication year, language, subject area, or DOI (when provided).
    try to keep the search queries short and simple, and not too specific (consider ambiguities).`,
  'Semantic Scholar': `
    next, generate a few valid Semantic Scholar search queries, based on the questions and prompt, by concatenating with "+" a few keywords.
    try to use various search strategies in the multiple queries. for example, if asked to compare topics A and B, you could search for A+B,
    and you could also search for A or B in separate queries and then compare the results.
    only if explicitly required in the prompt, you can use additional fields to filter the results, such as &year=, &publicationTypes=, &fieldsOfStudy=.
    keep the search queries short and simple.`,
};

export interface JarvisSettings {
  openai_api_key: string;
  scopus_api_key: string;
  springer_api_key: string;
  model: string;
  temperature: number;
  max_tokens: number;
  top_p: number;
  frequency_penalty: number;
  presence_penalty: number;
  timeout: number;
  paper_search_engine: string;
  use_wikipedia: boolean;
  paper_space: number;
  paper_tokens: number;
  include_paper_summary: boolean;
};

export function get_settings(): JarvisSettings {
  const settings = vscode.workspace.getConfiguration('jarvis');
  return {
    openai_api_key: settings.get('openAI.apiKey') as string,
    scopus_api_key: settings.get('scopus.apiKey') as string,
    springer_api_key: settings.get('springer.apiKey') as string,
    model: settings.get('openAI.model') as string,
    temperature: settings.get('openAI.temperature') as number,
    max_tokens: settings.get('openAI.maxTokens') as number,
    top_p: settings.get('openAI.topP') as number,
    frequency_penalty: settings.get('openAI.frequencyPenalty') as number,
    presence_penalty: settings.get('openAI.presencePenalty') as number,
    timeout: 1000*(settings.get('openAI.timeout') as number),
    paper_search_engine: settings.get('research.searchEngine') as string,
    use_wikipedia: settings.get('research.searchWikipedia') as boolean,
    paper_space: settings.get('research.paperSpace') as number,
    paper_tokens: settings.get('research.paperTokens') as number,
    include_paper_summary: settings.get('research.includeSummary') as boolean,
  };
}


src/extension.ts
import * as vscode from 'vscode';
import { do_research } from './research';

export function activate(context: vscode.ExtensionContext) {
	console.log('Jarvis is now active');

	let disposable = vscode.commands.registerCommand('jarvis.research', () => {
		// get selected text
		const editor = vscode.window.activeTextEditor;
		if (editor) {
			const selection = editor.selection;
			const text = editor.document.getText(selection);
      // replace selection with an empty string
      editor.edit(editBuilder => {
        editBuilder.replace(selection, '');
      });
      try {
        do_research(text);
      } catch (error) {
        vscode.window.showErrorMessage('Jarvis encountered an error. Try again.');
        console.log(error);
      }
	  }
	});

	context.subscriptions.push(disposable);
}

export function deactivate() {}


src/wikipedia.ts
/* eslint-disable @typescript-eslint/naming-convention */
import * as vscode from 'vscode';
import axios from 'axios';
import { query_completion } from './openai';
import { JarvisSettings } from './settings';
import { SearchParams } from './papers';
import { replace_selection } from './utils';

// interface with a string index signature
export interface WikiInfo {
  [key: string]: any;
  title?: string;
  year?: number;
  id?: number;
  excerpt?: string;
  text?: string;
  summary: string;
}

// return a summary of the top relevant wikipedia page
export async function search_wikipedia(prompt: string, search: SearchParams, settings: JarvisSettings): Promise<WikiInfo> {
  const search_term = await get_wikipedia_search_query(prompt, settings);
  if ( !search_term ) { return { summary: '' }; }

  const url = `https://en.wikipedia.org/w/api.php?action=query&list=search&origin=*&format=json&srlimit=20&srsearch=${search_term}`;
  const options = {
    headers: {'Accept': 'application/json'},
  };
  let response = await axios.get(url, options).catch((error) => { return error; });

  if (response.status !== 200) {
    console.log('wikipedia error ', response);
    return { summary: '' };
  }
  if (!response.data) { return { summary: '' }; }

  let pages: Promise<WikiInfo>[] = [];
  const jsonResponse = response.data;
  const results = jsonResponse['query']['search'];
  for (let i = 0; i < results.length; i++) {
    if (!results[i]['pageid']) { continue; }
    let page: WikiInfo = {
      title: results[i]['title'],
      year: parseInt(results[i]['timestamp'].split('-')[0]),
      id: results[i]['pageid'],
      excerpt: '',
      text: '',
      summary: '',
    };
    pages.push(get_wikipedia_page(page, 'excerpt', 'exintro'));
  }

  let best_page = await get_best_page(pages, results.length, search, settings);
  console.log('best page selected');
  best_page = await get_wikipedia_page(best_page, 'text', 'explaintext');
  console.log('full page text retrieved');
  best_page = await get_page_summary(best_page, search.questions, settings);
  console.log('page summary retrieved');
  return best_page;
}

async function get_wikipedia_search_query(prompt: string, settings: JarvisSettings): Promise<string> {
  const response = await query_completion(
    `define the main topic of the prompt.
    PROMPT:\n${prompt}
    use the following format.
    TOPIC: [main topic]`, settings);

  try {
    return response.split(/TOPIC:/gi)[1].replace(/"/g, '').trim();
  } catch {
    console.log(`bad wikipedia search query:\n${response}`);
    return '';
  }
}

// get the full text (or other extract) of a wikipedia page
async function get_wikipedia_page(page: WikiInfo, field:string = 'text', section: string = 'explaintext'): Promise<WikiInfo> {
  if ( !page['id'] ) { return page; }
  const url = `https://en.wikipedia.org/w/api.php?action=query&prop=extracts&${section}&format=json&pageids=${page['id']}`;
  const options = {
    headers: {'Accept': 'application/json'},
  };
  let response = await axios.get(url, options).catch((error) => { return error; });

  if (response.status !== 200) {
    console.log('wikipedia error ', response);
    return page;
  }
  if (!response.data || response.data.hasOwnProperty('error')) { return page; }

  const jsonResponse = response.data;
  const info = jsonResponse['query']['pages'][page['id'] as number];
  page[field] = info['extract'].replace(/<[^>]*>/g, '').trim();  // remove html tags

  return page;
}

async function get_best_page(pages: Promise<WikiInfo>[], n: number, search: SearchParams, settings: JarvisSettings): Promise<WikiInfo> {
  // TODO: we could do this by comparing 2 pages each time and keeping the max, at the cost of more queries
  let prompt = `you are a helpful assistant doing a literature review.
    we are searching for the single most relevant Wikipedia page to introduce the topics discussed in the research questions below.
    return only the index of the most relevant page in the format: [index number].
    QUESTIONS:\n${search.questions}\n
    PAGES:\n`;
  for (let i = 0; i < n; i++) {
    const page = await pages[i];
    if ( ! page['excerpt'] ) { continue; }
    if ( prompt.length + page['excerpt'].length > 0.9*4*settings.max_tokens ) {
      console.log(`stopping at ${i+1} pages due to max_tokens`);
      break;
    }

    prompt += `${i}. ${page['title']}: ${page['excerpt']}\n\n`;
  }
  const response = await query_completion(prompt, settings);
  const index = response.match(/\d+/);
  if (index) {
    return await pages[parseInt(index[0])];
  } else {
    return { summary: '' };
  }
}

async function get_page_summary(page: WikiInfo, questions: string, settings: JarvisSettings): Promise<WikiInfo> {
  if ( (!page['text']) || (page['text'].length === 0) ) { return page; }

  const user_p = settings.top_p;
  settings.top_p = 0.2;  // make the model more focused

  const prompt = 
    `here is a section from an article, research questions, and a draft summary of the complete article.
    if the section is not relevant to answering these questions, return the original summary unchanged in the response.
    otherwise, add to the summary information from the section that is relevant to the questions,
    and output the revised summary in the response.
    in the response, do not remove any relevant information that already exists in the summary,
    and describe how the article as a whole answers the given questions.`;

  let summary = 'empty summary.';
  const summary_step = 0.75*4*settings.max_tokens;
  for (let i=0; i<page['text'].length; i+=summary_step) {
    const text = page['text'].slice(i, i+summary_step);
    summary = await query_completion(
      `${prompt}
       SECTION: ${text}
       QUESTIONS: ${questions}
       SUMMARY: ${summary}
       RESPONSE:`, settings);
  }
  const decision = await query_completion(
    `decide if the following summary is relevant to any of the research questions below.
    only if it is not relevant to any of them, return "NOT RELEVANT", and explain why.
    SUMMARY:\n${summary}
    QUESTIONS:\n${questions}`,
    settings);
  console.log(decision);

  settings.top_p = user_p;

  if ((decision.includes('NOT RELEVANT')) || (summary.trim().length === 0)) {
    return page;
  }

  page['summary'] = `(Wikipedia, ${page['year']}) ${summary.replace(/\n+/g, ' ')}`;
  const wikilink = page['title'] ? page['title'].replace(/ /g, '_') : '';
  let cite = `- Wikipedia, [${page['title']}](https://en.wikipedia.org/wiki/${wikilink}), ${page['year']}.\n`;
  if (settings.include_paper_summary) {
    cite += `\t- ${page['summary']}\n`;
  }
  replace_selection(cite);

  return page;
}


src/papers.ts
/* eslint-disable @typescript-eslint/naming-convention */
import * as vscode from 'vscode';
import axios from 'axios';
import { AxiosResponse } from 'axios';
import { query_completion } from './openai';
import { JarvisSettings, search_prompts } from './settings';
import { replace_selection } from './utils';

export interface PaperInfo {
  title: string;
  author: string;
  year: number;
  journal: string;
  doi: string;
  citation_count: number;
  text: string;
  summary: string;
  compression: number;
};

export interface SearchParams {
  prompt: string;
  response: string;
  queries: string[];
  questions: string;
};

export async function search_papers(prompt: string, n: number, settings: JarvisSettings,
    min_results: number = 10, retries: number = 2): Promise<[PaperInfo[], SearchParams]> {

  const search = await get_search_queries(prompt, settings);
  console.log('search_papers');

  // run multiple queries in parallel and remove duplicates
  let results: PaperInfo[] = [];
  let dois: Set<string> = new Set();
  (await Promise.all(
    search.queries.map((query) => {
      if ( settings.paper_search_engine === 'Scopus' ) {
        return run_scopus_query(query, n, settings);
      } else if ( settings.paper_search_engine === 'Semantic Scholar' ) {
        return run_semantic_scholar_query(query, n);
      }
    })
    )).forEach((query) => {
      if ( !query ) { return; }
      query.forEach((paper) => {
        if (!dois.has(paper.doi)) {
          results.push(paper);
          dois.add(paper.doi);
        }
    });
  });

  if ( (results.length < min_results) && (retries > 0) ) {
    console.log(`search ${retries - 1}`);
    return search_papers(prompt, n, settings, min_results, retries - 1);
  }
  return [results, search];
}

async function run_semantic_scholar_query(query: string, papers: number): Promise<PaperInfo[]> {
  const options = {
    headers:{ 'Accept': 'application/json' },
  };

  // calculates the number of pages needed to fetch n results
  let limit = Math.min(papers, 100);
  let pages = Math.ceil(papers / limit);

  let start = 0;
  let results: PaperInfo[] = [];

  for (let p = 0; p < pages; p++) {
    const url = `https://api.semanticscholar.org/graph/v1/paper/search?query=${query}&limit=${limit}&page=${start}&fields=abstract,authors,title,year,venue,journal,citationCount,externalIds`;
    let response = await axios.get(url, options).catch((error) => { return error; });

    if ( response.status === 429 ) {
      console.log(response);
      vscode.window.showErrorMessage('Semantic Scholar API rate limit exceeded');
      continue;
    } else if ( response.status !== 200 ) {
      console.log(response);
      vscode.window.showErrorMessage('Scopus API error');
      continue;
    }

    const data = response.data;
    let papers: any[] = [];
    if (data) {
      papers = data['data'];
    }
    if ( !data ) {
      start += 25;
      continue;
    }

    try {
      for (let i = 0; i < papers.length; i++) {
        let journal: string = papers[i]['venue'];
        if ( !journal ) {
          if ( papers[i]['journal'] ) {
            journal = papers[i]['journal']['name'];
          } else { journal = 'Unknown'; }
        }
        let author = 'Unknown';
        if ( papers[i]['authors'][0] ) {
          author = papers[i]['authors'][0]['name'].split(' ').slice(1).join(' ');  // last name
        }

        const info: PaperInfo = {
          title: papers[i]['title'],
          author: author,
          year: parseInt(papers[i]['year'], 10),
          journal: journal ? journal : 'Unknown',
          doi: papers[i]['externalIds']['DOI'],
          citation_count: papers[i]['citationCount'],
          text: papers[i]['abstract'],
          summary: '',
          compression: 1,
        };
        results.push(info);
      }
    } catch (error) {
      console.log(error);
    }
  }

  return results.slice(0, papers);
}

async function run_scopus_query(query: string, papers: number,
    settings: JarvisSettings): Promise<PaperInfo[]> {
  if ( !settings.scopus_api_key ) {
    vscode.window.showErrorMessage('Scopus API key not set');
    throw new Error('Scopus API key not set');
  }
  const headers = {
    'Accept': 'application/json',
    'X-ELS-APIKey': settings.scopus_api_key,
  };
  const options = {
    headers: headers,
  };

  // calculates the number of pages needed to fetch n results
  let pages = Math.ceil(papers / 25);

  let start = 0;
  let results: PaperInfo[] = [];

  for (let p = 0; p < pages; p++) {
    const url = `https://api.elsevier.com/content/search/scopus?query=${query}&count=25&start=${start}&sort=-relevancy,-citedby-count,-pubyear`;
    let response = await axios.get(url, options).catch((error) => { return error; });

    if ( response.status === 429 ) {
      vscode.window.showErrorMessage('Scopus API key rate limit exceeded');
      throw new Error('Scopus API key rate limit exceeded');
    } else if ( response.status === 401 ) {
      vscode.window.showErrorMessage('Scopus API key invalid');
      throw new Error('Scopus API key invalid');
    } else if ( response.status !== 200 ) {
      console.log(response);
      vscode.window.showErrorMessage('Scopus API error');
      continue;
    }

    const data = response.data;
    let papers: any[] = [];
    if (data) {
      papers = data['search-results']['entry'];
    }
    if (!data || !papers || papers[0].hasOwnProperty('error')) {
      start += 25;
      continue;
    }

    try {
      for (let i = 0; i < papers.length; i++) {
        try {
          const info: PaperInfo = {
            title: papers[i]['dc:title'],
            author: papers[i]['dc:creator'].split(', ')[0].split(' ')[0],
            year: parseInt(papers[i]['prism:coverDate'].split('-')[0], 10),
            journal: papers[i]['prism:publicationName'],
            doi: papers[i]['prism:doi'],
            citation_count: parseInt(papers[i]['citedby-count'], 10),
            text: papers[i]['dc:description'],
            summary: '',
            compression: 1,
          };
          results.push(info);
        } catch {
          console.log('skipped', papers[i]);
          continue;
        }
      }

      start += 25;
      if ( data['search-results']['opensearch:totalResults'] < start ) {
        break;
      }

    } catch (error) {
      console.log(error);
    }
  }

  return results.slice(0, papers);
}

async function get_search_queries(prompt: string, settings: JarvisSettings):
    Promise<SearchParams> {
  console.log('get_search_queries');
  const response = await query_completion(
    `you are writing an academic text.
    first, list a few research questions that arise from the prompt below.
    ${search_prompts[settings.paper_search_engine]}
    PROMPT:\n${prompt}
    use the following format for the response.
    # [Title of the paper]

    ## Research questions

    1. [main question]
    2. [secondary question]
    3. [additional question]

    ## Queries

    1. [search query]
    2. [search query]
    3. [search query]
    `, settings);

  const query = response.split(/# Research questions|# Queries/gi);

  return {
    prompt: prompt,
    response: response.trim().replace(/## Research questions/gi, '## Prompt\n\n' + prompt + '\n\n## Research questions') + '\n\n## References\n\n',
    queries: query[2].trim().split('\n').map((q) => { return q.substring(q.indexOf(' ') + 1); }),
    questions: query[1].trim()
  };
}

export async function sample_and_summarize_papers(papers: PaperInfo[], max_tokens: number,
    search: SearchParams, settings: JarvisSettings): Promise<PaperInfo[]> {
  console.log('start sample_and_summarize_papers');
  let results: PaperInfo[] = [];
  let tokens = 0;

  // randomize the order of the papers
  papers.sort(() => Math.random() - 0.5);
  const promises: Promise<PaperInfo>[] = [];
  for (let i = 0; i < papers.length; i++) {

    if ( promises.length <= i ) {
      // try to get a summary for the next 5 papers asynchonously
      for (let j = 0; j < 5; j++) {
        if (i + j < papers.length) {
          promises.push(get_paper_summary(papers[i + j], search.questions, settings));
        }
      }
    }
    // wait for the next summary to be ready
    papers[i] = await promises[i];
    if ( papers[i]['summary'].length === 0 ) { continue; }

    // we only summarize papers up to a total length of max_tokens
    if (tokens + papers[i]['summary'].length / 4 > max_tokens) { 
      break;
    }
    results.push(papers[i]);
    tokens += papers[i]['summary'].length / 4;
  }

  console.log(`sampled ${results.length} papers. retrieved ${promises.length} papers.`);
  return results;
}

async function get_paper_summary(paper: PaperInfo,
    questions: string, settings: JarvisSettings): Promise<PaperInfo> {
  paper = await get_paper_text(paper, settings);
  if ( !paper['text'] ) { return paper; }

  const user_temp = settings.temperature;
  settings.temperature = 0.3;
  const prompt = `you are a helpful assistant doing a literature review.
    if the study below contains any information that pertains to topics discussed in the research questions below,
    return a summary in a single paragraph of the relevant parts of the study.
    only if the study is completely unrelated, even broadly, to these questions,
    return: 'NOT RELEVANT.' and explain why it is not helpful.
    QUESTIONS:\n${questions}
    STUDY:\n${paper['text']}`;
  const response = await query_completion(prompt, settings);
  //  consider the study's aim, hypotheses, methods / procedures, results / outcomes, limitations and implications.
  settings.temperature = user_temp;

  if (response.includes('NOT RELEVANT') || (response.trim().length === 0)) {
    paper['summary'] = '';
    return paper;
  }

  paper['summary'] = `(${paper['author']}, ${paper['year']}) ${response.replace(/\n+/g, ' ')}`;
  paper['compression'] = paper['summary'].length / paper['text'].length;

  let cite = `- ${paper['author']} et al., [${paper['title']}](https://doi.org/${paper['doi']}), ${paper['journal']}, ${paper['year']}, cited: ${paper['citation_count']}.\n`;
  if (settings.include_paper_summary) {
    cite += `\t- ${paper['summary']}\n`;
  }

  // replace selection in the editor with the citation
  replace_selection(cite);

  return paper;
}

async function get_paper_text(paper: PaperInfo, settings: JarvisSettings): Promise<PaperInfo> {
  if (paper['text']) { return paper; }  // already have the text
  let info = await get_scidir_info(paper, settings);  // ScienceDirect (Elsevier), full text or abstract
  if (info['text']) { return info; }
  else {
    info = await get_semantic_scholar_info(paper, settings);  // Semantic Scholar, abstract
    if (info['text']) { return info; }
    else {
      info = await get_crossref_info(paper);  // Crossref, abstract
      if (info['text']) { return info; }
      else {
        info = await get_springer_info(paper, settings);  // Springer, abstract
        if (info['text']) { return info; }
        else {
          return await get_scopus_info(paper, settings);  // Scopus, abstract
        }
      }
    }
  }
}

async function get_crossref_info(paper: PaperInfo): Promise<PaperInfo> {
  const url = `https://api.crossref.org/works/${paper['doi']}`;
  const headers = {
    "Accept": "application/json",
  };
  const options = {
    timeout: 5000,
    headers: headers,
  };
  const response = await axios.get(url, options).catch((error) => { return error; });

  if (response.status !== 200) {
    console.log('crossref error: ', response);
    return paper;
  }

  const info = response.data['message'];
  if ( info.hasOwnProperty('abstract') && (typeof info['abstract'] === 'string') ) {
    console.log('crossref success!');
    paper['text'] = info['abstract'].trim();
  }
  return paper;
}

async function get_scidir_info(paper: PaperInfo, settings: JarvisSettings): Promise<PaperInfo> {
  if (!settings.scopus_api_key) { return paper; }

  const url = `https://api.elsevier.com/content/article/doi/${paper['doi']}`;
  const headers = {
    'Accept': 'application/json',
    'X-ELS-APIKey': settings.scopus_api_key,
  };
  const options = {
    timeout: 5000,
    headers: headers,
  };
  const response = await axios.get(url, options).catch((error) => { return error; });

  if (response.status !== 200) {
    console.log('science_direct error: ', response);
    return paper;
  }

  const info = response.data['full-text-retrieval-response'];
  if ( (info['originalText']) && (typeof info['originalText'] === 'string') ) {

    try {
      const regex = new RegExp(/Discussion|Conclusion/gmi);
      if (regex.test(info['originalText'])) {
        // get end of main text
        paper['text'] = info['originalText']
          .split(/\bReferences/gmi).slice(-2)[0]
          .split(/Acknowledgments|Acknowledgements/gmi).slice(-2)[0]
          .split(regex).slice(-1)[0];

      } else {
        // get start of main text
        paper['text'] = info['originalText'].split(/http/gmi)[-1];  // remove preceding urls
      }
      paper['text'] = paper['text'].slice(0, 0.75*4*settings.max_tokens).trim();
      console.log('science_direct success (full)!');
    } catch {
      paper['text'] = '';
      console.log('science_direct: error parsing full text.');
    }
  }
  if ( !paper['text'] && info['coredata']['dc:description'] ) {
    paper['text'] = info['coredata']['dc:description'].trim();
    console.log('science_direct success (abstract)!');
  }
  return paper;
}

async function get_scopus_info(paper: PaperInfo, settings: JarvisSettings): Promise<PaperInfo> {
  if (!settings.scopus_api_key) { return paper; }

  const url = `https://api.elsevier.com/content/abstract/doi/${paper['doi']}`;
  const headers = {
    'Accept': 'application/json',
    'X-ELS-APIKey': settings.scopus_api_key,
  };
  const options = {
    timeout: 5000,
    headers: headers,
  };
  const response = await axios.get(url, options).catch((error) => { return error; });

  if (response.status !== 200) {
    console.log('scopus error: ', response);
    return paper;
  }

  const info = response.data['abstracts-retrieval-response']['coredata'];
  if ( info['dc:description'] ) {
    paper['text'] = info['dc:description'].trim();
    console.log('scopus success!');
  }
  return paper;
}

async function get_springer_info(paper: PaperInfo, settings: JarvisSettings): Promise<PaperInfo> {
  if ( !settings.springer_api_key ) { return paper; }

  const url = `https://api.springernature.com/metadata/json/doi/${paper['doi']}?api_key=${settings.springer_api_key}`;
  const headers = {
    'Accept': 'application/json',
  };
  const options = {
    timeout: 5000,
    headers: headers,
  };
  const response = await axios.get(url, options).catch((error) => { return error; });

  if (response.status !== 200) {
    console.log('springer: ', response);
    return paper;
  }

  if (response.data['records'].length === 0) { return paper; }
  const info = response.data['records'][0]['abstract'];
  if ( info ) {
    console.log('springer success!');
      paper['text'] = info.trim();
    }
  return paper;
}

async function get_semantic_scholar_info(paper: PaperInfo, settings: JarvisSettings): Promise<PaperInfo> {
  const url = `https://api.semanticscholar.org/v1/paper/DOI:${paper['doi']}?fields=abstract`;
  const headers = {
    'Accept': 'application/json',
  };
  const options = {
    timeout: 5000,
    headers: headers,
  };
  const response = await axios.get(url, options).catch((error) => { return error; });

  if (response.status !== 200) {
    console.log('semantic_scholar: ', response);
    return paper;
  }

  const info = response.data['abstract'];
  if ( info ) {
    console.log('semantic_scholar success!');
      paper['text'] = info.trim();
    }
  return paper;
}
